src/daft-table/src/probe_table/mod.rs

```diff
@@ -156,11 +163,188 @@ impl ProbeTable {
+
+                            let l_table = self.tables.get(l_table_idx).unwrap();
+
+                            let left_refs = l_table.0.as_slice();
+
+                            (self.compare_fn)(left_refs, &right_arrays, l_row_idx, r_idx).is_eq()
+                        }
+                    })
+                    .is_some()
+                {
+                    (r_idx as u64, true)
+                } else {
+                    (r_idx as u64, false)
+                }
+            }
+            None => (r_idx as u64, false),
+        }))
+    }
+
+    fn add_table(&mut self, table: &Table) -> DaftResult<()> {
+        // we have to cast to the join key schema
+        assert_eq!(table.schema, self.schema);
+        let hashes = table.hash_rows()?;
+        let table_idx = self.tables.len();
+        let table_offset = table_idx << Self::TABLE_IDX_SHIFT;
+
+        assert!(table_idx < (1 << (64 - Self::TABLE_IDX_SHIFT)));
+        assert!(table.len() < (1 << Self::TABLE_IDX_SHIFT));
+        let current_arrays = table
+            .columns
+            .iter()
+            .map(|s| Ok(s.as_physical()?.to_arrow()))
+            .collect::<DaftResult<Vec<_>>>()?;
+        self.tables.push(ArrowTableEntry(current_arrays));
+        let current_array_refs = self.tables.last().unwrap().0.as_slice();
+        for (i, h) in hashes.as_arrow().values_iter().enumerate() {
+            let idx = table_offset | i;
+            let entry = self.hash_table.raw_entry_mut().from_hash(*h, |other| {
+                (*h == other.hash) && {
+                    let j_idx = other.idx;
+                    let j_table_idx = (j_idx >> Self::TABLE_IDX_SHIFT) as usize;
+                    let j_row_idx = (j_idx & Self::LOWER_MASK) as usize;
+
+                    if table_idx == j_table_idx {
+                        (self.compare_fn)(current_array_refs, current_array_refs, i, j_row_idx)
+                            .is_eq()
+                    } else {
+                        let j_table = self.tables.get(j_table_idx).unwrap();
+
+                        let array_refs = j_table.0.as_slice();
+
+                        (self.compare_fn)(current_array_refs, array_refs, i, j_row_idx).is_eq()
+                    }
+                }
+            });
+            match entry {
+                RawEntryMut::Vacant(entry) => {
+                    entry.insert_hashed_nocheck(
+                        *h,
+                        IndexHash {
+                            idx: idx as u64,
+                            hash: *h,
+                        },
+                        (),
+                    );
+                    self.num_groups += 1;
+                }
+                RawEntryMut::Occupied(_) => {}
+            }
+        }
+        self.num_rows += table.len();
+        Ok(())
+    }
+}
+
+pub enum ProbeTable {
+    WithIdx(ProbeTableWithIdx),
+    WithoutIdx(ProbeTableWithoutIdx),
+}
```

Comment:
@samster25 Gonna keep thinking on this but wanted to ask if you have any opinions on how to allow for probe tables with different values? Right now I have this as an enum at the probe table level, but it's not the best as there's now quite a bit of duplicate code, and on the prober side it opens up more room for error as it now has to get the appropriate probe table type based on the join type.

Reply:
Yeah I'm not a fan of this either. Here's what i'm thinking as an alternative.
Since the API for building `add_table` will be the same, we can center that into a trait.

```
fn make_probe_table_builder(schema: SchemaRef, track_indices: bool) -> Box<dyn ProbeTableBuilder> {}

pub trait ProbeTableBuilder {
  fn add_table(&mut self, table: &Table) -> DaftResult<()>;

  fn build(&mut self) -> Box<dyn Probeable>;
}


// Both methods should work if we are tracking indices but if we disabled it, the later should only work.
pub trait Probeable {
  fn probe_indices(&self, table: &Table) -> Iterator<...>;

  fn probe_exists(&self, table: &Table) -> Iterator<bool>;
}
```